File: src/server/pdfChunker.ts
import { PDFParse } from "pdf-parse";

export interface ExtractedChunkRaw {
  chunkIndex: number;
  text: string;
  pageNumber: number;
  charLength: number;
  tokenEstimate: number;
}

const SENTENCE_BREAK_RE = /[.!?\n]\s/;

function splitStringWithOverlap(text: string, chunkSize: number, chunkOverlap: number): string[] {
  const result: string[] = [];
  if (!text) return result;

  const effectiveStep = Math.max(1, chunkSize - chunkOverlap);
  let startIndex = 0;

  while (startIndex < text.length) {
    let endIndex = startIndex + chunkSize;

    if (endIndex < text.length) {
      // Look for natural sentence break (. ! ? \n) near the end boundary
      const windowStart = Math.max(startIndex, endIndex - 80);
      const lookbackWindow = text.slice(windowStart, endIndex);
      const match = lookbackWindow.search(SENTENCE_BREAK_RE);

      if (match !== -1) {
        const offsetFromStart = windowStart + match + 1;
        if (offsetFromStart > startIndex + 50) {
          endIndex = offsetFromStart;
        }
      }
    } else {
      endIndex = text.length;
    }

    const chunkStr = text.slice(startIndex, endIndex).trim();
    if (chunkStr) {
      result.push(chunkStr);
    }

    if (endIndex >= text.length) {
      break;
    }
    startIndex += effectiveStep;
  }

  return result;
}

export function chunkTextContent(
  fullText: string,
  chunkSize: number,
  chunkOverlap: number,
  pagesText?: Array<{ pageNumber: number; text: string }> | null
): ExtractedChunkRaw[] {
  const chunks: ExtractedChunkRaw[] = [];

  if (pagesText && pagesText.length > 0) {
    let globalIndex = 0;
    for (const pageObj of pagesText) {
      const pageText = pageObj.text.trim();
      if (!pageText) continue;

      const pageChunks = splitStringWithOverlap(pageText, chunkSize, chunkOverlap);
      for (const textChunk of pageChunks) {
        const stripped = textChunk.trim();
        if (stripped.length > 10) {
          chunks.push({
            chunkIndex: globalIndex,
            text: stripped,
            pageNumber: pageObj.pageNumber,
            charLength: stripped.length,
            tokenEstimate: Math.ceil(stripped.length / 4),
          });
          globalIndex++;
        }
      }
    }
  } else {
    // Fallback if pages aren't structured individually
    const rawChunks = splitStringWithOverlap(fullText, chunkSize, chunkOverlap);
    let globalIndex = 0;
    for (const textChunk of rawChunks) {
      const stripped = textChunk.trim();
      if (stripped.length > 10) {
        // Estimate page number assuming ~2500 chars per page
        const estimatedPage = Math.floor((globalIndex * (chunkSize - chunkOverlap)) / 2500) + 1;
        chunks.push({
          chunkIndex: globalIndex,
          text: stripped,
          pageNumber: estimatedPage,
          charLength: stripped.length,
          tokenEstimate: Math.ceil(stripped.length / 4),
        });
        globalIndex++;
      }
    }
  }

  return chunks;
}

export async function parsePdfBuffer(buffer: Buffer): Promise<{
  text: string;
  pages: Array<{ pageNumber: number; text: string }>;
  pageCount: number;
}> {
  const parser = new PDFParse(new Uint8Array(buffer));
  const parsed = await parser.getText();
  const pages = (parsed.pages || []).map((p: any) => ({
    pageNumber: p.num,
    text: p.text || "",
  }));

  return {
    text: parsed.text || pages.map((p: any) => p.text).join("\n\n"),
    pages,
    pageCount: parsed.total || pages.length || 1,
  };
}

File: backend/pdf_chunker.py
"""PDF Processing & Text Chunking Engine (ported from src/lib/pdfChunker.ts)."""

import math
import re
from dataclasses import dataclass
from typing import Optional


@dataclass
class ExtractedChunkRaw:
    chunkIndex: int
    text: str
    pageNumber: int
    charLength: int
    tokenEstimate: int


_SENTENCE_BREAK_RE = re.compile(r"[.!?\n]\s")


def _split_string_with_overlap(text: str, chunk_size: int, chunk_overlap: int) -> list[str]:
    """Splits a string into chunks with overlap, preserving sentence boundaries where possible."""
    result: list[str] = []
    if not text:
        return result

    effective_step = max(1, chunk_size - chunk_overlap)
    start_index = 0

    while start_index < len(text):
        end_index = start_index + chunk_size

        if end_index < len(text):
            # Look for natural sentence break (. ! ? \n) near the end boundary
            window_start = max(start_index, end_index - 80)
            lookback_window = text[window_start:end_index]
            match = _SENTENCE_BREAK_RE.search(lookback_window)

            if match is not None:
                offset_from_start = window_start + match.start() + 1
                if offset_from_start > start_index + 50:
                    end_index = offset_from_start
        else:
            end_index = len(text)

        chunk_str = text[start_index:end_index].strip()
        if chunk_str:
            result.append(chunk_str)

        if end_index >= len(text):
            break
        start_index += effective_step

    return result


def chunk_text_content(
    full_text: str,
    chunk_size: int,
    chunk_overlap: int,
    pages_text: Optional[list[dict]] = None,
) -> list[ExtractedChunkRaw]:
    """Splits extracted text into structured chunks with page numbers and overlap."""
    chunks: list[ExtractedChunkRaw] = []

    if pages_text:
        global_index = 0
        for page_obj in pages_text:
            page_text = page_obj["text"].strip()
            if not page_text:
                continue

            page_chunks = _split_string_with_overlap(page_text, chunk_size, chunk_overlap)
            for text_chunk in page_chunks:
                stripped = text_chunk.strip()
                if len(stripped) > 10:
                    chunks.append(
                        ExtractedChunkRaw(
                            chunkIndex=global_index,
                            text=stripped,
                            pageNumber=page_obj["pageNumber"],
                            charLength=len(stripped),
                            tokenEstimate=math.ceil(len(stripped) / 4),
                        )
                    )
                    global_index += 1
    else:
        # Fallback if pages aren't structured individually
        raw_chunks = _split_string_with_overlap(full_text, chunk_size, chunk_overlap)
        global_index = 0
        for text_chunk in raw_chunks:
            stripped = text_chunk.strip()
            if len(stripped) > 10:
                # Estimate page number assuming ~2500 chars per page
                estimated_page = math.floor((global_index * (chunk_size - chunk_overlap)) / 2500) + 1
                chunks.append(
                    ExtractedChunkRaw(
                        chunkIndex=global_index,
                        text=stripped,
                        pageNumber=estimated_page,
                        charLength=len(stripped),
                        tokenEstimate=math.ceil(len(stripped) / 4),
                    )
                )
                global_index += 1

    return chunks
